org.eclipse.m2e.core/src/org/eclipse/m2e/core/embedder/IMaven.java

```diff
@@ -149,11 +152,42 @@ MojoExecution setupMojoExecution(MavenProject project, MojoExecution execution,
       throws CoreException;
 
   /**
+   * Resolves a configuration parameter from the given {@code mojoExecution}. It coerces from String to the given type
+   * and considers expressions and default values.
+   * 
+   * @param <T>
+   * @param project the Maven project
+   * @param mojoExecution the mojo execution from which to retrieve the configuration value
+   * @param parameter the name of the parameter (may be nested with separating {@code .})
+   * @param asType the type to coerce to
+   * @param monitor the progress monitor
+   * @return the parameter value or {@code null} if the parameter with the given name was not found
+   * @throws CoreException
    * @since 1.4
+   * @see IMaven#getMojoParameterValue(MavenProject, MojoExecution, List, Class, IProgressMonitor)
    */
   <T> T getMojoParameterValue(MavenProject project, MojoExecution mojoExecution, String parameter,
       Class<T> asType, IProgressMonitor monitor) throws CoreException;
 
+  /**
+   * Resolves a nested configuration parameter from the given {@code mojoExecution}. It coerces from String to the given
+   * type and considers expressions and default values.
+   * 
+   * @param <T>
+   * @param project the Maven project
+   * @param mojoExecution the mojo execution from which to retrieve the configuration value
+   * @param parameterPath the path of the parameter to look up, the first item is the name of the element directly below
+   *          {@code <configuration>} and the last one is the element containing the actual value
+   * @param asType the type to coerce to
+   * @param monitor the progress monitor
+   * @return the parameter value or {@code null} if the parameter with the given name was not found
+   * @throws CoreException
+   * @since 2.1.0
+   * @see IMaven#getMojoParameterValue(MavenProject, MojoExecution, String, Class, IProgressMonitor)
+   */
+  <T> T getMojoParameterValue(MavenProject project, MojoExecution mojoExecution, List<String> parameterPath,
```

Comment:
I assume that actually one would like to get an object of the 'real' configuration field type, but that is often not reasonably doable for complex types because they are in a Maven plugin associated library that one actually does not want to import.
In this case it would be the `EnforcerRule` class.

Therefore I wonder if instead of specifying a parameter path, it would be reasonable to define some kind of `ConfigurationProxy` that that is then returned and a caller can query for (child) elements:
```
public interface ConfigurationElementProxy {
  ConfigurationElementProxy get(String childName);
  <T> T as(Class<T>type);
}
```
With that the returned value can be treated more similar like an instance of the 'real' configuration field type.
In the case of this PR one could then do something like:
```
var rules = maven.getMojoParameterValue(mavenProject, mojoExecution, "rules", ConfigurationProxy.class, monitor)
String version = rules.get("requireJavaVersion").get("version").as(String.class)
```

The returned proxy would then just have to keep a reference to the instances needed to perform the final conversion in `as()`.

What do you and the others think?

Reply:
This method only allows to return value objects (https://maven.apache.org/guides/mini/guide-configuring-plugins.html#mapping-value-objects). For retrieving beans it is not suited! That requires much more refactoring and I currently don't see a use case for that. Your proposal would require to load the intermediate beans which is not necessary in this proposal here.

Reply:
I'm not sure if we have the same idea. Created https://github.com/eclipse-m2e/m2e-core/pull/996 as draft to demonstrate it.
Comments are welcome. :)

Reply:
@HannesWell Any idea how to move forward here given that there was no consensus for #996 yet?

Reply:
> I am also fine with removing the API exposure of the new method in IMaven and just use the implementation directly. Then there is time afterwards to decide on the proper API and migrate this usage to it.

Yes please continue that way. Changing internales later is no problem.

Can you please also add a Test case for the issue?

Reply:
Just one note, please don't add "convenience" methods to `IMaven` this methods do not add any value in terms of object instance and cann be static methods on a ConfigurationHelper or alike or you could use the MavenToolBox instead that already has things like session so one do not needs to pass it as a parameter directly.

So please what ever is choose here take the opportunity to move the code out of the `MavenImpl` to a more suitable place.

Reply:
Fixed in https://github.com/eclipse-m2e/m2e-core/pull/981/commits/36e1c4a4c3759ad35394fa097fa1afd60ba2a6c5